### ISSUE TITLE
S1751 False positive: delegate returning value wrongly considered as affecting loop

### ISSUE THREAD

**Issue by SebastienLorrain** [SonarSource/sonar-dotnet#2441]
**Repro steps**

```
public delegate T FalsePositive1_Delegate<out T>();

/// <summary>
/// The return in a delegate is wrongly considered as a return in the owner loop.
/// </summary>
private static void FalsePositive()
{
	List<object> entries = new List<object>() { new object() };

	foreach (object entry in entries)
	{
		FalsePositive_Sub<int>(
			delegate () { return -1; }  // Return in delegate is considered as return in the foreach.
			);
	}

	// Went past the loop.
	Console.WriteLine("Went past the loop");        //NOSONAR.
}

private static void FalsePositive_Sub<T>(FalsePositive1_Delegate<T> action)
{
	// Nothing to do.
}
```

**Expected behavior**

Rule S1751 should not be triggered

**Actual Behavior**

The return in the delegate is wrongly considered as early exit of the loop

**Related information**

Sonar C# version  7.11

**Comment by christophe-zurn-sonarsource** [comment]
Hello @SebastienLorrain,
Thanks for your feedback. Indeed, I confirm this is a FP on our side.

Rule S1751 incorrectly raise an issue when a `return` statement is found inside an inline delegate function using C# 2 syntax.

```
class Test
    {
        delegate int Delegate1();
        void myFunc()
        {
            while (true)
            {
                Delegate1 d2 = () => { return -1; }; // This is OK
            }
        }

        void myFunc2()
        {
            while (true)
            {
                Delegate1 d1 = delegate () { return -1; }; // Noncompliant FP
            }
        }
    }
```
